openedx/frontend-build
# Feat: reduce build time with cache

Most MFEs are using @edx/frontend-build to build their static files for both production and development configurations. However, this setup only works effectively for production because it ignores certain Webpack plugins that are unnecessary in a production environment.

This proposed change aims to enhance the build time for production by optimizing the Webpack cache configuration. It results in a significant improvement, with build times becoming more than 80% faster.

Issue related: https://github.com/overhangio/tutor-mfe/issues/88




**No cache**
![webpack-build-no-cache](https://github.com/openedx/frontend-build/assets/134975835/bdb05084-a8a7-4c01-bb2f-919903c50055) 

**Cached**
  ![webpack-build-cache](https://github.com/openedx/frontend-build/assets/134975835/c65d9a61-9911-474f-b592-677b0be26547)



<table>
  <tr>
    <th colspan="3">No cache</th>
    <th colspan="3">Cached</th>
  </tr>
  <tr>
    <th>ms</th>
    <th>seconds</th>
    <th>minutes</th>
    <th>ms</th>
    <th>seconds</th>
    <th>minutes</th>
  </tr>
  <tr>
    <td>109849</td>
    <td>109,849</td>
    <td>1,83</td>
    <td>3988</td>
    <td>3,988</td>
    <td>0,06</td>
  </tr>
</table>

----

**Note:** We may need to add a volume in tutor to the .cache folder for each MFE and introduce an additional configuration to enable or disable this feature within tutor
Certainly! Here's the corrected version of the text:

#### Example of mfe authn previous to the change

![mfe-auth-no-cached](https://github.com/openedx/frontend-build/assets/134975835/1d17fa92-4bbd-458a-99b5-9a17b6e31f7a)

### How to test it:

1. Clone the repository using the `feat-build-time` branch:
   ```
   git clone -b feat-build-time https://github.com/johnvente/frontend-build.git
   ```

2. Install the necessary dependencies:
   ```
   npm install
   ```

3. Clone a Micro Frontend (MFE) inside the `frontend-build` folder that builds its assets using `@edx/frontend-build`. This means that the `webpack.prod.config.js` file of the MFE should have the following configuration:

   ```javascript
   const { createConfig } = require('@edx/frontend-build');
   ```

   Example MFE repositories with this requirement:
   - https://github.com/openedx/frontend-app-learner-dashboard
   - https://github.com/openedx/frontend-app-gradebook
   - https://github.com/openedx/frontend-app-learner-dashboard
   - https://github.com/openedx/frontend-app-authn.git
  
  
4. Add this variable to `.env `  file:

  ```shell
   ENABLE_WEBPACK_CACHE=''  
   ```

5. Run `npm install` inside each cloned MFE.

6. Update the `webpack.prod.config.js` file of the cloned MFE(s) to use the following import statement:

   ```javascript
   const { createConfig } = require('../index');
   ```

7. Run the build command:
   ```
   npm run build
   ```

   Note: The first time you run the build command, you may not notice any difference because the `.cache` folder has not been created yet.

8. Run the build command again:
   ```
   npm run build
   ```

Please make sure to review and adapt the instructions based on your specific setup and requirements.

## Note:
To implement this feature we need to add `.cache` folder to the mfe that will be using this and add the following to the ` .env ` file 

  ```
   ENABLE_WEBPACK_CACHE=''
   ```
